arrows/vxl/high_pass_filter.cxx

```diff
@@ -0,0 +1,489 @@
+    vil_math_image_abs_difference( grey_img, filter_xy, filter_xy );
+
+    return output;
+  }
+
+  // Given an input grayscale image, and a smoothed version of this greyscale
+  // image, calculate the bidirectional filter response in the horizontal
+  // direction.
+  template < typename PixType >
+  void
+  horizontal_box_bidirectional_pass( const vil_image_view< PixType >& grey,
+                                     const vil_image_view< PixType >& smoothed,
+                                     vil_image_view< PixType >& output,
+                                     unsigned kernel_width )
+  {
+    const unsigned ni = grey.ni();
+    const unsigned nj = grey.nj();
+
+    output.fill( 0 );
+
+    unsigned offset = ( kernel_width / 2 ) + 1;
+
+    unsigned diff1 = 0, diff2 = 0;
+
+    if( ni < 2 * offset + 1 )
+    {
+      return;
+    }
+
+    for( unsigned j = 0; j < nj; j++ )
+    {
+      for( unsigned i = offset; i < ni - offset; i++ )
+      {
+        const PixType& val = grey( i, j );
+        const PixType& avg1 = smoothed( i - offset, j );
+        const PixType& avg2 = smoothed( i + offset, j );
+
+        diff1 = ( val > avg1 ? val - avg1 : avg1 - val );
+        diff2 = ( val > avg2 ? val - avg2 : avg2 - val );
+
+        output( i, j ) = std::min( diff1, diff2 );
+      }
+    }
+  }
+
+  // Given an input grayscale image, and a smoothed version of this greyscale
+  // image, calculate the bidirectional filter response in the vertical
+  // direction.
+  template < typename PixType >
+  void
+  vertical_box_bidirectional_pass( const vil_image_view< PixType >& grey,
+                                   const vil_image_view< PixType >& smoothed,
+                                   vil_image_view< PixType >& output,
+                                   unsigned kernel_height )
+  {
+    vil_image_view< PixType > grey_t = vil_transpose( grey );
+    vil_image_view< PixType > smoothed_t = vil_transpose( smoothed );
+    vil_image_view< PixType > output_t = vil_transpose( output );
+
+    horizontal_box_bidirectional_pass( grey_t,
+                                       smoothed_t,
+                                       output_t,
+                                       kernel_height );
+  }
+
+  // Perform bidirectional filtering
+  template < typename PixType >
+  vil_image_view< PixType >
+  bidirection_box_filter( const vil_image_view< PixType >& grey_img )
+  {
+    vil_image_view< PixType > output;
+    output.set_size( grey_img.ni(), grey_img.nj(), 3 );
+
+    // Create views of each plane
+    vil_image_view< PixType > filter_x = vil_plane( output, 0 );
+    vil_image_view< PixType > filter_y = vil_plane( output, 1 );
+    vil_image_view< PixType > filter_xy = vil_plane( output, 2 );
+
+    // Report the difference between the pixel value, and all of the smoothed
+    // responses. The xy channel is used as a temporary buffer to avoid
+    // additional memory allocation
```

Comment:
```suggestion
    // Report the difference between the pixel value, and all of the smoothed
    // responses. The xy channel is used as a temporary buffer to avoid
    // additional memory allocation.
```

...or...

```suggestion
    // Report the difference between the pixel value, and all of the smoothed
    // responses, using the xy channel as a temporary buffer to avoid
    // additional memory allocation
```